src/basic/terminal-util.c

```diff
@@ -2436,6 +2436,129 @@ int terminal_get_size_by_dsr(
         return r;
 }
 
+#define MAX_TERMINFO_LENGTH 64
+/* python -c 'print("".join(hex(ord(i))[2:] for i in "TN").upper())' */
+#define DCS_TERMINFO_Q ANSI_DCS "+q" "544E" ANSI_ST
+/* The answer is either 0+r… (invalid) or 1+r… (OK). */
+#define DCS_TERMINFO_R0 ANSI_DCS "0+r" "544E" ANSI_ST
+#define DCS_TERMINFO_R1 ANSI_DCS "1+r" "544E" "=" /* This is followed by Pt ST. */
+
+static int scan_terminfo_response(
+                const char *buf,
+                size_t size,
+                char **ret_name) {
+        int r;
+
+        assert(buf);
+        assert(ret_name);
+
+        /* Check if the have enough space for an answer */
+        if (size < STRLEN(DCS_TERMINFO_R0))
+                return 0;
+
+        /* Check if the terminating sequence is present */
+        if (memcmp(buf + size - STRLEN(ANSI_ST), ANSI_ST, STRLEN(ANSI_ST)) != 0)
+                return 0;
+
+        if (size <= STRLEN(DCS_TERMINFO_R1 ANSI_ST))
+                return -EINVAL;  /* The answer is invalid or empty */
+
+        if (memcmp(buf, DCS_TERMINFO_R1, STRLEN(DCS_TERMINFO_R1)) != 0)
+                return -EINVAL;  /* The answer is not valid */
+
+        _cleanup_free_ void *dec = NULL;
+        size_t dec_size;
+        r = unhexmem_full(buf + STRLEN(DCS_TERMINFO_R1), size - STRLEN(DCS_TERMINFO_R1 ANSI_ST),
+                          /* secure= */ false,
+                          &dec, &dec_size);
+        if (r < 0)
+                return r;
+
+        if (strlen(dec) != dec_size || string_has_cc(dec, NULL))
```

Comment:
hmm, the left check only is safe because unhexmem() is nice to you and appends a hidden NUL byte implicitly. but that's a safey net only. would prefer if this check was changed to `memchr(dec, 0, dec_size)`. which makes this safe.